### PR TITLE
APPS-1308 Update Apply config endpoint

### PR DESCRIPTION
**Pull request by korotkov-aerospike**
1) on app startup moved instantiation on top level, reducing number of params
2) configuration manager return model.Config on read
3) make configuration manager impls private
4) (most important) on each CRUD operation call changeConfig method, that updates config, saves it and applies.
5) on reschedule leave ad-hoc jobs

### CODE REVIEW

File: cmd/backup/main.go [hunk 1]
```diff
@@ -75,24 +75,47 @@ func startService(configFile string, remote bool) error {
...
+	err = configApplier.ApplyNewConfig()
 	if err != nil {
 		return err
 	}
 
 	var restoreJobs = service.NewRestoreJobsHolder()
-	service.NewMetricsCollector(handlers, restoreJobs).Start(ctx, 1*time.Second)
+	service.NewMetricsCollector(backupHandlers, restoreJobs).Start(ctx, 1*time.Second)
+
+	restoreMgr := service.NewRestoreManager(backends, config, service.NewRestoreGo(), clientManager, restoreJobs)
```
Comment (reugn) [inline cmd/backup/main.go]: Consider renaming `NewRestoreGo` -> `NewRestore`

File: internal/server/handlers/config.go [hunk 2]
```diff
@@ -51,59 +61,87 @@ func (s *Service) updateConfig(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	err = s.configurationManager.Write(ctx, s.config)
+	if err != nil {
+		return fmt.Errorf("failed to write configuration: %w", err)
 	}
+
+	err = s.configApplier.ApplyNewConfig()
+	if err != nil {
+		return fmt.Errorf("failed to apply new configuration:  %w", err)
```
Comment (reugn) [inline internal/server/handlers/config.go]:
```suggestion
		return fmt.Errorf("failed to apply new configuration: %w", err)
```

File: internal/server/handlers/config.go [hunk 3]
```diff
@@ -51,59 +61,87 @@ func (s *Service) updateConfig(w http.ResponseWriter, r *http.Request) {
...
+	config, err := s.configurationManager.Read(r.Context())
+	if err != nil {
+		hLogger.Error("failed to read config",
+			slog.Any("error", err),
+		)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	err = s.changeConfig(r.Context(), func(c *model.Config) error {
```
Comment (reugn) [inline internal/server/handlers/config.go]: `changeConfig` calls to `configurationManager.Write` which is unnecessary and also not supported for configuration managers other than local.

File: internal/server/handlers/config_policy.go [hunk 4]
```diff
@@ -39,9 +40,9 @@ func (s *Service) ConfigPolicyActionHandler(w http.ResponseWriter, r *http.Reque
 func (s *Service) addPolicy(w http.ResponseWriter, r *http.Request) {
 	hLogger := s.logger.With(slog.String("handler", "addPolicy"))
 
-	request, err := dto.NewBackupPolicyFromReader(r.Body, dto.JSON)
+	newPolicy, err := dto.NewBackupPolicyFromReader(r.Body, dto.JSON)
 	if err != nil {
-		hLogger.Error("failed to decode request body",
+		hLogger.Error("failed to decode newPolicy body",
```
Comment (reugn) [inline internal/server/handlers/config_policy.go]: Maybe just: failed to decode policy?

File: internal/server/handlers/config_policy.go [hunk 5]
```diff
@@ -166,9 +161,9 @@ func (s *Service) readPolicy(w http.ResponseWriter, r *http.Request) {
 func (s *Service) updatePolicy(w http.ResponseWriter, r *http.Request) {
 	hLogger := s.logger.With(slog.String("handler", "updatePolicy"))
 
-	request, err := dto.NewBackupPolicyFromReader(r.Body, dto.JSON)
+	updatedPolicy, err := dto.NewBackupPolicyFromReader(r.Body, dto.JSON)
 	if err != nil {
-		hLogger.Error("failed to decode request body",
+		hLogger.Error("failed to decode newPolicyDto body",
```
Comment (reugn) [inline internal/server/handlers/config_policy.go]: failed to decode policy

File: internal/server/handlers/service.go [hunk 6]
```diff
@@ -10,34 +11,35 @@ import (
 )
 
 type Service struct {
+	sync.Mutex
```
Comment (reugn) [inline internal/server/handlers/service.go:14]: It seems to be not in use.

File: pkg/service/backup_backends_holder.go [hunk 7]
```diff
@@ -9,23 +9,30 @@ import (
...
 type BackendsHolder interface {
+	// Init creates new backends from config.
+	Init(config *model.Config)
 	// GetReader returns BackupBackend for routine as BackupListReader.
 	GetReader(routineName string) (BackupListReader, bool)
 	// Get returns BackupBackend for routine.
 	Get(routineName string) (*BackupBackend, bool)
-	// SetData replaces stored backends.
-	SetData(backends map[string]*BackupBackend)
+	// GetAllReaders returns all backends as BackupListReader.
```
Comment (reugn) [inline pkg/service/backup_backends_holder.go]: as a map of

File: pkg/service/backup_scheduler.go [hunk 8]
```diff
@@ -11,9 +11,15 @@ import (
 	"github.com/reugn/go-quartz/quartz"
 )
 
+type quartzGroup = string
+type jobType string
```
Comment (reugn) [inline pkg/service/backup_scheduler.go]:
```suggestion
type (
	quartzGroup string
	jobType     string
)
```

File: pkg/service/config_applier.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// we don't want to delete ad-hoc jobs
+func (a *DefaultConfigApplier) clearPeriodicSchedulerJobs() error {
+	keys, err := a.scheduler.GetJobKeys(matcher.JobGroupEquals(quartzGroupScheduled))
```
Comment (reugn) [inline pkg/service/config_applier.go]:
```suggestion
	keys, err := a.scheduler.GetJobKeys(matcher.JobGroupEquals(string(quartzGroupScheduled)))
```

File: pkg/service/config_applier.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+	if err != nil {
+		return fmt.Errorf("cannot fetch jobs: %w", err)
+	}
+
+	slog.Info(fmt.Sprintf("Delete scheduled jobs %+v", keys))
```
Comment (reugn) [inline pkg/service/config_applier.go]:
```suggestion
	slog.Info("Delete scheduled jobs", slog.Any("keys", keys))
```

File: internal/server/configuration/configuration_manager.go [hunk 11]
```diff
@@ -15,11 +15,10 @@ import (
 )
 
 type Manager interface {
-	ReadConfiguration(ctx context.Context) (io.ReadCloser, error)
-	WriteConfiguration(ctx context.Context, config *model.Config) error
+	Read(ctx context.Context) (*model.Config, error)
+	Write(ctx context.Context, config *model.Config) error
```
Comment (reugn) [inline internal/server/configuration/configuration_manager.go]: Add documentation for the interface methods.

File: internal/server/handlers/config.go [hunk 12]
```diff
@@ -1,13 +1,24 @@
...
-	"github.com/aerospike/aerospike-backup-service/v2/pkg/service"
+	"github.com/aerospike/aerospike-backup-service/v2/pkg/model"
 )
 
+func (s *Service) ConfigActionHandler(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case http.MethodGet:
+		s.readConfig(w)
+	case http.MethodPut:
+		s.updateConfig(w, r)
```
Comment (reugn) [inline internal/server/handlers/config.go:18]: Add the default option returning `http.StatusNotFound` or `http.StatusMethodNotAllowed`.

**Review by reugn — COMMENTED** [review]
Does the refactoring require regenerating OpenAPI spec?

File: internal/server/handlers/service.go (hunk 6 above)
Comment (korotkov-aerospike) [reply internal/server/handlers/service.go:14]: used in changeConfig

**Review by reugn — APPROVED** [review]

**Pull request merged**